src/docbuilder/queue.rs

```diff
@@ -87,6 +87,12 @@ impl DocBuilder {
                 .map(|r| PackageKind::Registry(r.as_str()))
                 .unwrap_or(PackageKind::CratesIo);
 
+            if let Err(err) = builder.update_toolchain() {
+                log::error!("Updating toolchain failed, locking queue: {}", err);
+                self.lock()?;
+                return Err(err);
+            }
```

Comment:
Rather than calling `update_toolchain` in more places, do you think it makes sense to change `build_package` to return an error enum so we can tell where the error came from? The current change seems racy: a new nightly could be published between calling update_toolchain here and calling it in `build_package`.

I'm also slightly concerned that this doesn't handle errors when an admin runs `cratesfyi build crate`, but I don't know a simple way to handle that, and it seems rare for that to be the first build with a new toolchain anyway.

Reply:
That was the approach I tried first, but I got stuck with working out how to make it work with `failure`, I could take another attempt.

Reply:
Hmm, could you maybe add a new error type and try downcasting to that? Don't spend too much time on it :) this should work fine 99.9% of the time.